src/crimson/osd/pg.cc

```diff
@@ -847,22 +864,24 @@ std::optional<hobject_t> PG::resolve_oid(
   const SnapSet &ss,
   const hobject_t &oid)
 {
+  logger().debug("{} oid.snap={},head snapset.seq={}",
+                 __func__, oid.snap, ss.seq);
   if (oid.snap > ss.seq) {
+    // perfrom a read from head
     return oid.get_head();
   } else {
     // which clone would it be?
-    auto clone = std::upper_bound(
+    auto clone = std::lower_bound(
       begin(ss.clones), end(ss.clones),
       oid.snap);
     if (clone == end(ss.clones)) {
       // Doesn't exist, > last clone, < ss.seq
       return std::nullopt;
     }
     auto citer = ss.clone_snaps.find(*clone);
-    // TODO: how do we want to handle this kind of logic error?
-    ceph_assert(citer != ss.clone_snaps.end());
```

Comment:
If citer == ss.clone_snaps.end(), we have an entry in ss.clones without a corresponding ss.clone_snaps entry.  That is meant to be impossible.  For now, an assert is fine.  In the future, we'd probably want to propagate some kind of logic error so scrub can try to fix it.